src/report_reader.cpp

```diff
@@ -423,7 +438,8 @@ DataFrame<T> ReportReader<T>::Population::get(const nonstd::optional<Selection>&
         off_t offset = 0;
         off_t data_offset = (timer_index - index_start) / stride;
         auto data_ptr = &data_frame.data[data_offset * n_ids];
-        for (const auto& position : positions) {
+        for (const auto& node_pointer : node_pointers) {
```

Comment:
when the full selection is used, we know that, and can avoid the full copy of the `node_pointers` map, which can be quite large, no?

Reply:
As I explained in my previous comments, this PR does not change the behaviour of the code. The code used to copy all the node IDs into a vector, and then all the positions into a separate vector. The memory consumption increases in the current code.

The PR simplifies this by just using the same data structure instead.

Reply:
Can't we make the improvement here though?

Reply:
I'm going to give it a go; I will report back what I get.

Reply:
I tried switching to a map here:
https://github.com/BlueBrain/libsonata/compare/sandbox/srivas/ids_perf_improvements...sandbox/gevaert/ids-perf-improvements?expand=1

The case of fetching the full report is faster, so I think this should be pursued:
```
Map:
#IDS    | 1         | 2         | 3
      1 | 0.021234  | 0.004357  | 0.004069
     10 | 5.125821  | 4.981242  | 4.990284
    100 | 5.579101  | 5.563343  | 5.612758
   1000 | 5.994568  | 6.000875  | 6.436316
  10000 | 6.057319  | 6.065421  | 6.588673
 100000 | 6.722486  | 6.660799  | 7.208886
1000000 | 13.149290 | 13.744862 | 14.119598
4234929 | 37.245648 | 39.671034 | 39.629977

Vector:
#IDS    | 1          | 2         | 3
      1 | 0.005059   | 0.003865  | 0.003995
     10 | 5.104329   | 4.995507  | 5.019744
    100 | 5.963634   | 5.681968  | 5.617718
   1000 | 6.389871   | 6.375495  | 6.052416
  10000 | 6.196432   | 6.461813  | 6.104821
 100000 | 6.767185   | 7.148655  | 6.833382
1000000 | 12.229618  | 12.654732 | 12.594303
4234929 | 30.915542  | 32.286852 | 31.678779
```
Each column is a run, where a run loads the # of random IDS on the left; it does this for 3 full time frames, and averages the read time of a frame.

Reply:
Thank you very much, @mgeplf. Indeed, the `std::map` is no longer necessary and switching to a vector will make the memory accesses considerably faster.

I just squashed your branch, cleaned it up a bit, and pushed the changes into this branch. I will now verify that no major changes are observed in performance compared to the old version, just for the shake of completeness.

Reply:
Here is the commit:
https://github.com/BlueBrain/libsonata/pull/174/commits/e1dce9d50264aacd97766d0b875708567aeeee58

Reply:
The same evaluation of [fetching one GID of the figure](https://github.com/BlueBrain/libsonata/pull/174#issuecomment-1018826442) but with the update to `std::vector`, that also illustrates the same improvements that @mgeplf mentioned for large amounts of GIDs:

```
1 / 0.010879278182983398
10 / 5.343438386917114
100 / 5.943392753601074
1000 / 6.406604766845703
10000 / 6.451685428619385
100000 / 7.101217269897461
1000000 / 12.762017250061035
4234929 / 27.868659019470215
```

Reply:
cool; thanks.